core/amber/src/main/python/core/util/expression_evaluator/__init__.py

```diff
@@ -0,0 +1,105 @@
+import re
+from typing import Any, Dict, List, Optional, Pattern
+
+from proto.edu.uci.ics.amber.engine.architecture.worker import EvaluatedValue, TypedValue
+
+
+class ExpressionEvaluator:
+    """
+    Provides a series of static evaluation methods of a given expression, with an optional context.
+    """
+
+    @staticmethod
+    def evaluate(expression: str, runtime_context: Optional[Dict[str, Any]] = None) -> EvaluatedValue:
+
+        value = eval(expression, runtime_context)
+        value_str = repr(value)
+        type_str = type(value).__name__
+
+        attributes = list()
+
+        # add attributes
+        attributes += ExpressionEvaluator._extract_attributes(value)
+
+        # add container items
+        attributes += ExpressionEvaluator._extract_container_items(value)
+        return EvaluatedValue(
+            value=TypedValue(
+                expression=expression,
+                value_ref=expression,
+                value_str=value_str,
+                value_type=type_str,
+                expandable=ExpressionEvaluator._is_expandable(value)),
+            attributes=attributes
+        )
+
+    @staticmethod
+    def _is_expandable(obj) -> bool:
+        return ExpressionEvaluator._contains_attributes(obj) or \
+               (ExpressionEvaluator._is_iterable(obj) and not ExpressionEvaluator._is_empty_container(obj))
+
+    @staticmethod
+    def _is_iterable(obj) -> bool:
+        return hasattr(obj, "__iter__")
+
+    @staticmethod
+    def _contains_attributes(obj) -> bool:
+        return hasattr(obj, "__dict__") and len(obj.__dict__) > 0
+
+    @staticmethod
+    def _is_empty_container(obj) -> bool:
+        return hasattr(obj, "__len__") and len(obj) == 0
+
+    @staticmethod
+    def _contextualize_expression(expression: str, context_replacements: Dict[Pattern[str], str]) -> str:
+
+        contextualized_expression = expression
+        for pattern, contextualized_pattern in context_replacements.items():
+            contextualized_expression = re.sub(pattern, contextualized_pattern, contextualized_expression)
+        return contextualized_expression
+
+    @staticmethod
+    def _extract_container_items(value: Any) -> List[TypedValue]:
+        contained_items = []
+        if ExpressionEvaluator._is_iterable(value):
```

Comment:
just wondering, how does this code behave for iterator objects? For example, there's a hand-written iterator object that just implements `__iter__` and `__next__`, how does our code deal with it?